### ISSUE TITLE
Any advise on how to introduce discrete changes to the learning rate?

### ISSUE THREAD

**Issue by slothkong** [kbullaughey/dni-synthetic-gradients#1]
Hello,  I would like to pass a particular learning rate to the optim algorithm, according to the number of iterations (3x10^-5 until reaching 300k [1250 epochs], 3x10^-6 until 400k [1666 epochs] and 3x10^-7 until  500[ [2084 epochs] ), this for the purpose of replicating the experiment in the DNIs paper.  
I changed SGD for Adam and tried this:
`adamState = adamState or {
        learningRate =( iter >= 1666 and 3e-7 or ( iter >= 1250 and 3e-6 or 3e-5)) ,  --[[other hyperparameters go here ]]}`

where `iter ` belongs to a for-loop which calls both the `tran()` and `test()` functions. However, upon running the code I get :  ./mnist-relu.lua:170: attempt to compare nil with number

Could it be that the `iter` counter is out of the scope of the `train()` function? This would be weird since `iter` is declared in the main body of the code (global variable?). Help?

**Comment by kbullaughey** [comment]
I suspect it must be some kind of scope problem, but I can't say much more without seeing the code. Do you have a fork of it on Github?

**Comment by slothkong** [comment]
Hi, I think I've created a fork now. I made the changes to the code a feed with the details is visible to me. But I'm not sure how if you can see them as well.

**Comment by slothkong** [comment]
Maybe this link will work: https://github.com/slothkong/dni-synthetic-gradients/commits/master/mnist-relu.lua 

First time trying to actually use GitHub features.

**Comment by kbullaughey** [comment]
Although most variables in lua are global scope unless explicitly declared local scope, one exception is that variables declared as the iteration variable of a for loop are [local scope to the for loop block](http://lua-users.org/wiki/ForTutorial). You could probably fix this by declaring iter outside the loop using a `while true` loop, manually incrementing iter, and breaking at the right iteration. This would mean you're using a globally scoped counter.

**Comment by slothkong** [comment]
I didn't know iterators where excepted form the global scope. Thanks for clearing that out.
